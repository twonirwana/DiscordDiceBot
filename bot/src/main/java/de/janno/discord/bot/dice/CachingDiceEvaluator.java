package de.janno.discord.bot.dice;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import de.janno.evaluator.dice.DiceEvaluator;
import de.janno.evaluator.dice.ExpressionException;
import de.janno.evaluator.dice.Roller;
import de.janno.evaluator.dice.random.NumberSupplier;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.Tags;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;

import java.util.concurrent.ExecutionException;

import static io.micrometer.core.instrument.Metrics.globalRegistry;

@Slf4j
public class CachingDiceEvaluator {

    private final LoadingCache<String, RollerOrError> diceRollerCache;
    private final DiceEvaluator diceEvaluator;

    public CachingDiceEvaluator(NumberSupplier numberSupplier, int maxNumberOfDice, int cacheSize) {
        diceEvaluator = new DiceEvaluator(numberSupplier, maxNumberOfDice);
        diceRollerCache = CacheBuilder.newBuilder()
                .maximumSize(cacheSize)
                .recordStats()
                .build(new CacheLoader<>() {
                    @Override
                    public @NonNull RollerOrError load(@NonNull String expression) {
                        try {
                            log.debug("create roller for: {}", expression);
                            Roller roller = diceEvaluator.buildRollSupplier(expression);
                            roller.roll();
                            return new RollerOrError(expression, roller, true, null);
                        } catch (ExpressionException | ArithmeticException e) {
                            return new RollerOrError(expression, null, false, e.getMessage());
                        }
                    }
                });
        Gauge.builder("diceEvaluator.cache", diceRollerCache::size).tags(Tags.of("stats", "size")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().requestCount()).tags(Tags.of("stats", "requests")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().hitCount()).tags(Tags.of("stats", "hit")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().missCount()).tags(Tags.of("stats", "miss")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().totalLoadTime()).tags(Tags.of("stats", "totalLoadTime")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().averageLoadPenalty()).tags(Tags.of("stats", "averageLoadTime")).register(globalRegistry);
    }

    public @NonNull RollerOrError get(@NonNull String expression) {
        try {
            return diceRollerCache.get(expression);
        } catch (ExecutionException e) {
            throw new IllegalStateException(e);
        }
    }

    @VisibleForTesting
    public long getCacheSize() {
        return diceRollerCache.size();
    }
}
