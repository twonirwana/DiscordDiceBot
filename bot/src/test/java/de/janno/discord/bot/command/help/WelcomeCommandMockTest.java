package de.janno.discord.bot.command.help;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import de.janno.discord.bot.ButtonEventAdaptorMock;
import de.janno.discord.bot.SlashEventAdaptorMock;
import de.janno.discord.bot.command.channelConfig.ChannelConfigCommand;
import de.janno.discord.bot.command.customDice.CustomDiceCommand;
import de.janno.discord.bot.command.customParameter.CustomParameterCommand;
import de.janno.discord.bot.command.sumCustomSet.SumCustomSetCommand;
import de.janno.discord.bot.dice.CachingDiceEvaluator;
import de.janno.discord.bot.persistance.PersistenceManager;
import de.janno.discord.bot.persistance.PersistenceManagerImpl;
import de.janno.discord.connector.api.slash.CommandInteractionOption;
import de.janno.evaluator.dice.random.RandomNumberSupplier;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicLong;

import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SnapshotExtension.class)
public class WelcomeCommandMockTest {


  PersistenceManager persistenceManager;
    Expect expect;

    @BeforeEach
    void setup() {
        persistenceManager = new PersistenceManagerImpl("jdbc:h2:mem:" + UUID.randomUUID(), null, null);
    }

    @Test
    void slashStartWelcome_firstButton() {
        WelcomeCommand underTest = getWelcomeCommand();

        SlashEventAdaptorMock slashEvent = new SlashEventAdaptorMock(List.of(CommandInteractionOption.builder()
                .name("start")
                .build()));
        underTest.handleSlashCommandEvent(slashEvent, () -> UUID.fromString("00000000-0000-0000-0000-000000000001"), Locale.ENGLISH).block();

        expect.scenario("slash").toMatchSnapshot(slashEvent.getSortedActions());

        Optional<ButtonEventAdaptorMock> buttonEvent = slashEvent.getFirstButtonEventMockOfLastButtonMessage();
        assertThat(buttonEvent).isPresent();
        underTest.handleComponentInteractEvent(buttonEvent.get()).block();
        expect.scenario("button").toMatchSnapshot(buttonEvent.get().getSortedActions());
    }

    @Test
    void slashStartWelcome_legacyDnD5Id() {
        WelcomeCommand underTest = getWelcomeCommand();

        SlashEventAdaptorMock slashEvent = new SlashEventAdaptorMock(List.of(CommandInteractionOption.builder()
                .name("start")
                .build()));
        underTest.handleSlashCommandEvent(slashEvent, () -> UUID.fromString("00000000-0000-0000-0000-000000000001"), Locale.ENGLISH).block();

        expect.scenario("slash").toMatchSnapshot(slashEvent.getSortedActions());

        ButtonEventAdaptorMock buttonEvent = new ButtonEventAdaptorMock("welcome", "dnd5_image", new AtomicLong(0));
        underTest.handleComponentInteractEvent(buttonEvent).block();
        expect.scenario("button").toMatchSnapshot(buttonEvent.getSortedActions());
    }

    @Test
    void slashStartWelcome_invalidId() {
        WelcomeCommand underTest = getWelcomeCommand();

        SlashEventAdaptorMock slashEvent = new SlashEventAdaptorMock(List.of(CommandInteractionOption.builder()
                .name("start")
                .build()));
        underTest.handleSlashCommandEvent(slashEvent, () -> UUID.fromString("00000000-0000-0000-0000-000000000001"), Locale.ENGLISH).block();

        expect.scenario("slash").toMatchSnapshot(slashEvent.getSortedActions());

        ButtonEventAdaptorMock buttonEvent = new ButtonEventAdaptorMock("welcome", "invalidId", new AtomicLong(0));
        underTest.handleComponentInteractEvent(buttonEvent).block();
        expect.scenario("button").toMatchSnapshot(buttonEvent.getSortedActions());
    }


    private WelcomeCommand getWelcomeCommand() {
        CachingDiceEvaluator cachingDiceEvaluator = new CachingDiceEvaluator(new RandomNumberSupplier(0));

        CustomDiceCommand customDiceCommand = new CustomDiceCommand(persistenceManager, cachingDiceEvaluator);
        CustomParameterCommand customParameterCommand = new CustomParameterCommand(persistenceManager, cachingDiceEvaluator);
        SumCustomSetCommand sumCustomSetCommand = new SumCustomSetCommand(persistenceManager, cachingDiceEvaluator);
        ChannelConfigCommand channelConfigCommand = new ChannelConfigCommand(persistenceManager);
        RpgSystemCommandPreset rpgSystemCommandPreset = new RpgSystemCommandPreset(persistenceManager, customParameterCommand, customDiceCommand, sumCustomSetCommand, channelConfigCommand);

        return new WelcomeCommand(persistenceManager, rpgSystemCommandPreset, () -> UUID.fromString("00000000-0000-0000-0000-000000000000"));
    }



}
