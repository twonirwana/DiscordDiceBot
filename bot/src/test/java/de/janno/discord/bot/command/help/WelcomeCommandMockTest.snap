de.janno.discord.bot.command.help.WelcomeCommandMockTest.slashStartWelcome_firstButton[button]=[
  [
    "editMessage: message:Welcome to the Button Dice Bot,\nuse one of the example buttons below to start one of the RPG dice systems, use `/quickstart system` to select one of many RPG presets or use the slash command to configure your own custom dice system (see https://github.com/twonirwana/DiscordDiceBot for details or the slash command `/help`).\nYou can also use the slash command `/r` to directly roll dice with.\nFor help or feature request come to the support discord server: https://discord.gg/e43BsqKpFr, buttons=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=Dungeon & Dragons 5e, id=welcome\u001EDND5_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dungeon & Dragons 5e without Dice Images, id=welcome\u001EDND5\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Fate, id=welcome\u001EFATE_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=ðŸª™Coin Toss, id=welcome\u001ECOIN\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=nWod / Chronicles of Darkness, id=welcome\u001ENWOD\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=oWod / Storyteller System, id=welcome\u001EOWOD\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Shadowrun, id=welcome\u001ESHADOWRUN_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dice Calculator, id=welcome\u001EDICE_CALCULATOR\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)])]",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=Click on a button to roll the dice, fields=[], componentRowDefinitions=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=D20, id=custom_dice\u001E1_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=D20 Advantage, id=custom_dice\u001E2_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=D20 Disadvantage, id=custom_dice\u001E3_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=1d4, id=custom_dice\u001E4_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=1d6, id=custom_dice\u001E5_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=1d8, id=custom_dice\u001E6_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=1d10, id=custom_dice\u001E7_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=1d12, id=custom_dice\u001E8_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=1d100, id=custom_dice\u001E9_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d4, id=custom_dice\u001E10_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d6, id=custom_dice\u001E11_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d8, id=custom_dice\u001E12_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d10, id=custom_dice\u001E13_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=2d12, id=custom_dice\u001E14_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d20, id=custom_dice\u001E15_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)])], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=`/custom_dice start buttons: D20;2d20k1@D20 Advantage;2d20L1@D20 Disadvantage;;1d4;1d6;1d8;1d10;1d12;1d100;2d4=@2d4;2d6=@2d6;2d8=@2d8;2d10=@2d10;2d12=@2d12;2d20=@2d20 answer_format: without_expression answer_interaction: none dice_image_style: polyhedral_RdD dice_image_color: default`, fields=[], componentRowDefinitions=[], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)"
  ]
]


de.janno.discord.bot.command.help.WelcomeCommandMockTest.slashStartWelcome_firstButton[slash]=[
  [
    "reply: commandString",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=Welcome to the Button Dice Bot,\nuse one of the example buttons below to start one of the RPG dice systems, use `/quickstart system` to select one of many RPG presets or use the slash command to configure your own custom dice system (see https://github.com/twonirwana/DiscordDiceBot for details or the slash command `/help`).\nYou can also use the slash command `/r` to directly roll dice with.\nFor help or feature request come to the support discord server: https://discord.gg/e43BsqKpFr, fields=[], componentRowDefinitions=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=Dungeon & Dragons 5e, id=welcome\u001EDND5_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dungeon & Dragons 5e without Dice Images, id=welcome\u001EDND5\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Fate, id=welcome\u001EFATE_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=ðŸª™Coin Toss, id=welcome\u001ECOIN\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=nWod / Chronicles of Darkness, id=welcome\u001ENWOD\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=oWod / Storyteller System, id=welcome\u001EOWOD\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Shadowrun, id=welcome\u001ESHADOWRUN_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dice Calculator, id=welcome\u001EDICE_CALCULATOR\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)])], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)"
  ]
]


de.janno.discord.bot.command.help.WelcomeCommandMockTest.slashStartWelcome_invalidId[button]=[
  [
    "editMessage: message:Welcome to the Button Dice Bot,\nuse one of the example buttons below to start one of the RPG dice systems, use `/quickstart system` to select one of many RPG presets or use the slash command to configure your own custom dice system (see https://github.com/twonirwana/DiscordDiceBot for details or the slash command `/help`).\nYou can also use the slash command `/r` to directly roll dice with.\nFor help or feature request come to the support discord server: https://discord.gg/e43BsqKpFr, buttons=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=Dungeon & Dragons 5e, id=welcome\u001EDND5_IMAGE\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dungeon & Dragons 5e without Dice Images, id=welcome\u001EDND5\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Fate, id=welcome\u001EFATE_IMAGE\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=ðŸª™Coin Toss, id=welcome\u001ECOIN\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=nWod / Chronicles of Darkness, id=welcome\u001ENWOD\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=oWod / Storyteller System, id=welcome\u001EOWOD\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Shadowrun, id=welcome\u001ESHADOWRUN_IMAGE\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dice Calculator, id=welcome\u001EDICE_CALCULATOR\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)])]"
  ]
]


de.janno.discord.bot.command.help.WelcomeCommandMockTest.slashStartWelcome_invalidId[slash]=[
  [
    "reply: commandString",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=Welcome to the Button Dice Bot,\nuse one of the example buttons below to start one of the RPG dice systems, use `/quickstart system` to select one of many RPG presets or use the slash command to configure your own custom dice system (see https://github.com/twonirwana/DiscordDiceBot for details or the slash command `/help`).\nYou can also use the slash command `/r` to directly roll dice with.\nFor help or feature request come to the support discord server: https://discord.gg/e43BsqKpFr, fields=[], componentRowDefinitions=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=Dungeon & Dragons 5e, id=welcome\u001EDND5_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dungeon & Dragons 5e without Dice Images, id=welcome\u001EDND5\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Fate, id=welcome\u001EFATE_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=ðŸª™Coin Toss, id=welcome\u001ECOIN\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=nWod / Chronicles of Darkness, id=welcome\u001ENWOD\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=oWod / Storyteller System, id=welcome\u001EOWOD\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Shadowrun, id=welcome\u001ESHADOWRUN_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dice Calculator, id=welcome\u001EDICE_CALCULATOR\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)])], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)"
  ]
]


de.janno.discord.bot.command.help.WelcomeCommandMockTest.slashStartWelcome_legacyDnD5Id[button]=[
  [
    "editMessage: message:Welcome to the Button Dice Bot,\nuse one of the example buttons below to start one of the RPG dice systems, use `/quickstart system` to select one of many RPG presets or use the slash command to configure your own custom dice system (see https://github.com/twonirwana/DiscordDiceBot for details or the slash command `/help`).\nYou can also use the slash command `/r` to directly roll dice with.\nFor help or feature request come to the support discord server: https://discord.gg/e43BsqKpFr, buttons=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=Dungeon & Dragons 5e, id=welcome\u001EDND5_IMAGE\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dungeon & Dragons 5e without Dice Images, id=welcome\u001EDND5\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Fate, id=welcome\u001EFATE_IMAGE\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=ðŸª™Coin Toss, id=welcome\u001ECOIN\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=nWod / Chronicles of Darkness, id=welcome\u001ENWOD\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=oWod / Storyteller System, id=welcome\u001EOWOD\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Shadowrun, id=welcome\u001ESHADOWRUN_IMAGE\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dice Calculator, id=welcome\u001EDICE_CALCULATOR\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)])]",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=Click on a button to roll the dice, fields=[], componentRowDefinitions=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=D20, id=custom_dice\u001E1_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=D20 Advantage, id=custom_dice\u001E2_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=D20 Disadvantage, id=custom_dice\u001E3_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=1d4, id=custom_dice\u001E4_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=1d6, id=custom_dice\u001E5_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=1d8, id=custom_dice\u001E6_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=1d10, id=custom_dice\u001E7_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=1d12, id=custom_dice\u001E8_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=1d100, id=custom_dice\u001E9_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d4, id=custom_dice\u001E10_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d6, id=custom_dice\u001E11_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d8, id=custom_dice\u001E12_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d10, id=custom_dice\u001E13_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=2d12, id=custom_dice\u001E14_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=2d20, id=custom_dice\u001E15_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false, emoji=null)])], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=`/custom_dice start buttons: D20;2d20k1@D20 Advantage;2d20L1@D20 Disadvantage;;1d4;1d6;1d8;1d10;1d12;1d100;2d4=@2d4;2d6=@2d6;2d8=@2d8;2d10=@2d10;2d12=@2d12;2d20=@2d20 answer_format: without_expression answer_interaction: none dice_image_style: polyhedral_RdD dice_image_color: default`, fields=[], componentRowDefinitions=[], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)"
  ]
]


de.janno.discord.bot.command.help.WelcomeCommandMockTest.slashStartWelcome_legacyDnD5Id[slash]=[
  [
    "reply: commandString",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=Welcome to the Button Dice Bot,\nuse one of the example buttons below to start one of the RPG dice systems, use `/quickstart system` to select one of many RPG presets or use the slash command to configure your own custom dice system (see https://github.com/twonirwana/DiscordDiceBot for details or the slash command `/help`).\nYou can also use the slash command `/r` to directly roll dice with.\nFor help or feature request come to the support discord server: https://discord.gg/e43BsqKpFr, fields=[], componentRowDefinitions=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=Dungeon & Dragons 5e, id=welcome\u001EDND5_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dungeon & Dragons 5e without Dice Images, id=welcome\u001EDND5\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Fate, id=welcome\u001EFATE_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=ðŸª™Coin Toss, id=welcome\u001ECOIN\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=nWod / Chronicles of Darkness, id=welcome\u001ENWOD\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=oWod / Storyteller System, id=welcome\u001EOWOD\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Shadowrun, id=welcome\u001ESHADOWRUN_IMAGE\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Dice Calculator, id=welcome\u001EDICE_CALCULATOR\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)])], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)"
  ]
]